### PR TITLE
Determine session type on stream based on stream watcher events

### PR DESCRIPTION
**Pull request by gabrielcorado**
Related to #49164

**Brief context:** After the changes are made at #47309, when a request for session recording events (playback) is made, we query to find the last event to set the `SessionType` field on the audit log. Depending on the cluster's backend and the volume of audit events, this can cause increased latency.

This PR updates the logic determining the session type for the `SessionRecordingAccess` audit event to receive a callback from the streamer when the first event is available.

changelog: Improve session playback initial delay caused by an additional events query.

### CODE REVIEW

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -196,30 +196,14 @@ func (a *ServerWithRoles) actionWithExtendedContext(kind, verb string, extendCon
 // actionForKindSession is a special checker that grants access to session
 // recordings. It can allow access to a specific recording based on the
 // `where` section of the user's access rule for kind `session`.
-func (a *ServerWithRoles) actionForKindSession(ctx context.Context, sid session.ID) (types.SessionKind, error) {
-	sessionEnd, err := a.findSessionEndEvent(ctx, sid)
-
-	extendContext := func(ctx *services.Context) error {
-		ctx.Session = sessionEnd
+func (a *ServerWithRoles) actionForKindSession(ctx context.Context, sid session.ID) error {
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/auth/auth_with_roles.go:199]: Do you mean we should update it as well? I reverted it to what it was before I changed it at #47309.

Looking at the usage of `ctx.Session`, it still requires to be a last event, but we could retrieve it using the new callback format.

[the remark this answers is not in the document]
Comment (zmb3) [reply lib/auth/auth_with_roles.go:199]: Ah, I guess it was always calling `findSessionEndEvent` - I thought that was new from your changes.

Disregard this thread - my mistake.

File: lib/events/auditlog.go [hunk 2]
```diff
@@ -545,8 +545,25 @@ func (l *AuditLog) StreamSessionEvents(ctx context.Context, sessionID session.ID
 		"session_id", string(sessionID),
 	)
 
+	sessionStartCh := make(chan apievents.AuditEvent, 1)
+	if watcher, err := EventsWatcherFromContext(ctx); err == nil {
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/events/auditlog.go]: Updated it. Now the error cases closes the channel (causing the callback to called if any).

**Comment by zmb3** [comment]
I haven't forgot about you @gabrielcorado - this is on my list to review/test later today!

File: lib/events/auditlog.go [hunk 3]
```diff
@@ -667,3 +691,39 @@ func (l *AuditLog) periodicSpaceMonitor() {
...
+
+// ContextWithSessionStartCallback returns a context.Context containing a
+// session start event callback.
+func ContextWithSessionStartCallback(ctx context.Context, cb SessionStartCallback) context.Context {
+	return context.WithValue(ctx, sessionStartCallbackContextKey, cb)
+}
+
+// SessionStartCallbackFromContext returns the session start callback from
+// context.Context.
+func SessionStartCallbackFromContext(ctx context.Context) (SessionStartCallback, error) {
```
Comment (zmb3) [inline lib/events/auditlog.go:718]: Does this need to be exported?

**Review by zmb3 — APPROVED** [review]

**Review by greedy52 — APPROVED** [review]

**Pull request merged**